tests/test_dependency_from_lifespan.py

```diff
@@ -0,0 +1,72 @@
+from contextlib import asynccontextmanager
+from typing import Any, AsyncIterator, Dict
+
+import pytest
+from fastapi import Depends, FastAPI, FromLifespan
+from fastapi.testclient import TestClient
+from typing_extensions import Annotated
+
+
+def test_app_extract_from_lifespan() -> None:
+    class DatabaseConnection:
+        def __init__(self, expected: int) -> None:
+            self.expected = expected
+
+        async def query(self, __query: str) -> Any:
+            assert __query == f"SELECT {self.expected}"
+            return self.expected
+
+    class DatabaseConnectionPool:
+        def __init__(self, expected: int) -> None:
+            self.expected = expected
+
+        @asynccontextmanager
+        async def acquire(self) -> AsyncIterator[DatabaseConnection]:
+            yield DatabaseConnection(self.expected)
+
+    @asynccontextmanager
+    async def connect(expected: int) -> AsyncIterator[DatabaseConnectionPool]:
+        yield DatabaseConnectionPool(expected)
+
+    @asynccontextmanager
+    async def lifespan(app: FastAPI) -> AsyncIterator[Dict[str, Any]]:
+        async with connect(42) as db:
+            yield {"db": db}
```

Comment:
I'm not a big fan of this design as a solution to lifespan dependency, it's really just syntactic sugar over the lifespan `state` like this.

The value of dependencies is the ability to inject dependencies in a more distributed way. For example I might have a lifespan dependency that's only applicable to one group of API routes so I want to define it as a lifespan dependency in that context rather than having to define it on a lifespan passed directly into the main FastAPI application.

Reply:
I think there’s actually quite a bit of value in having users manually wire the dependencies in the lifespan. It mirrors the “composition root” concept in most standalone dependency injection systems. It’s a singular well identified place where you tell the dependency injection system how to build your dependencies.

This keeps the execution model clear and simple. I’ve implemented the more “distributed” version in Xpresso and it can get messy.